File: 백준 N과 M (2) - 백트랙킹-/복습/Main.java
import java.util.Scanner;

public class Main {
    static boolean[] isVisited;
    static int[] num;
    static int n;
    static int m;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        isVisited = new boolean[m + 1];
        num = new int[m + 1];
        dfs(1, -1);
    }

    static void dfs(int current, int prevNum) {
        if (current > m) {
            for (int i = 1; i <= m; i++) {
                System.out.print(num[i] + " ");
            }
            System.out.println();
            return;
        }
        for (int i = 1; i <= n; i++) {
            if (isVisited[current]) {
                continue;
            }
            if (prevNum >= i) {
                continue;
            }
            isVisited[current] = true;
            num[current] = i;
            prevNum = i;
            dfs(current + 1, prevNum);
            isVisited[current] = false;
        }
    }
}

File: 백준 N과 M (2) - 백트랙킹-/복습/Main.kt
import java.util.*

internal var isVisited: BooleanArray? = null
internal var num: IntArray? = null
internal var n: Int = 0
internal var m: Int = 0

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    n = sc.nextInt()
    m = sc.nextInt()
    isVisited = BooleanArray(m + 1)
    num = IntArray(m + 1)
    dfs(1, -1)
}

internal fun dfs(current: Int, prevNum: Int) {
    var prevNum = prevNum
    if (current > m) {
        for (i in 1..m) {
            print(num!![i].toString() + " ")
        }
        println()
        return
    }
    for (i in 1..n) {
        if (isVisited!![current]) {
            continue
        }
        if (prevNum >= i) {
            continue
        }
        isVisited!![current] = true
        num?.set(current, i)
        prevNum = i
        dfs(current + 1, prevNum)
        isVisited!![current] = false
    }
}